pkg/source/ttn/source.go

```diff
@@ -0,0 +1,178 @@
+}
+
+// ExportDevice implements the source.Source interface.
+func (s *Source) ExportDevice(devID string) (*ttnpb.EndDevice, error) {
+	mgr, err := s.getDeviceManager(s.config.appID)
+	if err != nil {
+		return nil, err
+	}
+	dev, err := mgr.Get(devID)
+	if err != nil {
+		return nil, errors.FromGRPCError(err)
+	}
+
+	v3dev := &ttnpb.EndDevice{}
+	v3dev.DeviceID = dev.DevID
+	v3dev.ApplicationID = s.config.appID
+
+	v3dev.Name = dev.DevID
+	v3dev.Description = dev.Description
+	v3dev.Attributes = dev.Attributes
+
+	v3dev.JoinEUI = &types.EUI64{}
+	if err := v3dev.JoinEUI.Unmarshal(dev.AppEUI.Bytes()); err != nil {
+		return nil, err
+	}
+	v3dev.DevEUI = &types.EUI64{}
+	if err := v3dev.DevEUI.Unmarshal(dev.DevEUI.Bytes()); err != nil {
+		return nil, err
+	}
+
+	v3dev.LoRaWANVersion = ttnpb.MAC_V1_0_2
+	v3dev.LoRaWANPHYVersion = ttnpb.PHY_V1_0_2_REV_B
+	v3dev.FrequencyPlanID = s.config.frequencyPlanID
+
+	v3dev.MACSettings = &ttnpb.MACSettings{
+		Rx1Delay: &ttnpb.RxDelayValue{
+			Value: ttnpb.RX_DELAY_1,
+		},
+	}
+	if dev.Uses32BitFCnt {
+		v3dev.MACSettings.Supports32BitFCnt = &pbtypes.BoolValue{
+			Value: dev.Uses32BitFCnt,
+		}
+	}
+	if dev.DisableFCntCheck {
+		v3dev.MACSettings.ResetsFCnt = &pbtypes.BoolValue{
+			Value: dev.DisableFCntCheck,
+		}
+	}
+
+	if dev.AppKey != nil && !dev.AppKey.IsEmpty() {
+		v3dev.SupportsJoin = true
+		v3dev.RootKeys = &ttnpb.RootKeys{}
+		v3dev.RootKeys.AppKey = &ttnpb.KeyEnvelope{
+			Key: &types.AES128Key{},
+		}
+		if err := v3dev.RootKeys.AppKey.Key.Unmarshal(dev.AppKey.Bytes()); err != nil {
+			return nil, err
+		}
+	}
+
+	if dev.Latitude != 0 || dev.Longitude != 0 {
+		v3dev.Locations = map[string]*ttnpb.Location{
+			"user": {
+				Latitude:  float64(dev.Latitude),
+				Longitude: float64(dev.Longitude),
+				Altitude:  dev.Altitude,
+				Source:    ttnpb.SOURCE_REGISTRY,
+			},
+		}
+	}
+
+	if dev.DevAddr != nil && dev.NwkSKey != nil && dev.AppSKey != nil {
+		v3dev.Session = &ttnpb.Session{
+			SessionKeys: ttnpb.SessionKeys{
+				AppSKey: &ttnpb.KeyEnvelope{
+					Key: &types.AES128Key{},
+				},
+				FNwkSIntKey: &ttnpb.KeyEnvelope{
+					Key: &types.AES128Key{},
+				},
+			},
+			LastFCntUp:    dev.FCntUp,
+			LastNFCntDown: dev.FCntDown,
+			StartedAt:     time.Now(),
+		}
+		if err := v3dev.Session.DevAddr.Unmarshal(dev.DevAddr.Bytes()); err != nil {
+			return nil, err
+		}
+		if err := v3dev.Session.SessionKeys.AppSKey.Key.Unmarshal(dev.AppSKey.Bytes()); err != nil {
+			return nil, err
+		}
+		if err := v3dev.Session.SessionKeys.FNwkSIntKey.Key.Unmarshal(dev.NwkSKey.Bytes()); err != nil {
+			return nil, err
+		}
+	}
+
+	return v3dev, nil
+}
+
+// RangeDevices implements the source.Source interface.
+func (s *Source) RangeDevices(appID string, f func(source.Source, string) error) error {
+	mgr, err := s.getDeviceManager(appID)
+	if err != nil {
+		return err
+	}
+	devices, err := mgr.List(0, 0)
+	if err != nil {
+		return errors.FromGRPCError(err)
+	}
+
+	for _, dev := range devices {
```

Comment:
I certainly hope there is some delay somewhere else, otherwise you'll definitely run into rate limits when migrating a large number of devices. And since you're always starting from 0, it will be impossible to finish a migration.